Favicons (WIP)

Hinton opened the pull request:
🚧 This is an **early attempt** at adding favicon to the browser extension. 🚧 

I decided to open a pull request early in order to facilitate a discussion to ensure this is a good approach. See bitwarden/core#32 for the previous discussion on this topic.

UX wise this PR adds a favicon next to each login item. In the event a favicon does not exist it displays a default image.
![Screenshot](https://user-images.githubusercontent.com/137855/31252567-e2811cde-aa21-11e7-9b06-7af5769bbf31.png)

Implementation wise it uses an "img" tag to load the favicon directly from **Google**, which is far from an optimal solution. The next step would be to add a new "service" to the server part of Bitwarden to store and serve the favicons. Ideally the service should proxy and cache requests.

Another improvement which I would like to include is a local cache of the images, to minimize the required number of requests.

Privacy wise, this solution will leak the **hostname** of the login items to Bitwarden, which is why I added a setting to toggle the functionality. 🔒 

Grateful for feedback!

kspearrin commented:
Hey, thanks for the PR. This looks pretty good. I wonder if we should reverse the login of the settings so that is it disabled by default?

kspearrin commented:
Also, found this service which has the ability for us to spin up a service to support: https://icons.better-idea.org/

Hinton commented:
Nice find, unfortunately it seems like the service **redirects** the browser and loads the files directly from the websites themselves. Which is something I preferably want to avoid. The discussion over at bitwarden/core#32 discusses this approach more in detail, but the main disadvantage that was brought up is that the network administrators would be able to see requests going to every website.

What I would prefer to have is a service which finds favicons like besticon but would also download the icon and serve it itself. The network traffic thereafter would only be a couple of requests to icons.bitwarden.com.

I'm not against making it disabled by default, however it would probably hide the feature from the majority of users. Would it make sense to show a configuration page after login to enable/disable this feature and possibly Analytics? Or can we make it discoverable in some other way?

mhammond commented:
> but the main disadvantage that was brought up is that the network administrators would be able to see requests going to every website.

Firefox itself has a similar concern for its "Sync" product. When a device syncs (say) a bookmark or history entry to a new device, we don't drag the favicon with it. Users then complain the bookmark/visit has no favicon until the device actually visits the page. The primary reason we don't fetch them in this case is user privacy.

Hinton commented:
I know that Chrome provides a url scheme to retrieve the favicons from the browser cache, "chrome://favicon/" which could alleviate the privacy concerns. I've yet to find any information if the same is possible with Firefox with webextensions (and the other browsers Bitwarden supports). Although this would mean that only websites that have been visited would show a favicon.

kspearrin commented:
I see. I guess we would need to write some custom service that caches Google S2 or besticons on the backend to avoid that. Regardless, we also need to figure out a way to have these icon requests cached on the client. We can't have the client sending hundreds (and potentially thousands for some folks) requests for icons every time they load their vault listing.

Opt-out is fine I guess. We need to come up with some kind if configuration prompt in the future to help people understand these things a bit better, especially when it comes to privacy features.

kspearrin commented:
Additionally, if we were to serve these icons up from a cached CDN endpoint I suppose it would be ok not to have have to write any custom login to cache them on the client. Not sure how we would handle 404 cases when the icon doesn't exist though.

mhammond commented:
> I know that Chrome provides a url scheme to retrieve the favicons from the browser cache, "chrome://favicon/" which could alleviate the privacy concerns.

I agree - that would. I couldn't find a firefox bug for webext access to favicons, but if there's something that works like that in chrome I'd help with a bug for Firefox to support it. FWIW, I'm soon going on holidays for a month, so my help won't be timely...

kspearrin commented:
Since I imagine people will want this feature everywhere, the implementation must use a solution that can be accessible from all client applications, not just the browser extension.

Hinton commented:
I wrote a fairly quick "faviconService" which caches the favicons for 30 days, and stores them in the browser. The service needs some refactoring before I'm happy with it but early testings seems positive.

@kspearrin What version of EcmaScript is the project targeting? I noticed everything is written using the old syntax (ES5). Is it alright to target newer versions and use arrow functions, and async/await?

@mhammond On chrome I currently fall back to using "chrome://favicon/" if the user disabled the feature, however It's not currently working on Firefox sadly.

Hinton commented:
I updated the URL to icons.bitwarden.com.
There is a tiny delay between opening the vault view and the icons load from the cache, not sure if it can be improved somehow. I guess the div could have a fixed size to prevent the jumping.

kspearrin commented on src/popup/app/services/faviconService.js:
```diff
@@ -0,0 +1,119 @@
+
+        // Return the favicon for a specific uri.
+        _service.getFavicon = function (uri) {
+            var domain = utilsService.getDomain(uri);
```
I would suggest using `new URL()`'s `hostname` property here so that you get distinctions between subdomains. This will return `google.com` for `accounts.google.com` and `admin.google.com`.

Hinton replied:
Ah yea, that's probably a good idea. I wasn't sure if getDomain returned subdomains or just the top level.

kspearrin commented:
@Hinton best practice would be to load a default placeholder icon that is then asynchronously replaced once the proper icon is resolved. Many websites do this to give an illusion of things loading faster.

kspearrin commented on src/popup/app/services/faviconService.js:
```diff
@@ -69,15 +69,15 @@ angular
         // Return the favicon for a specific uri.
         _service.getFavicon = function (uri) {
-            var domain = utilsService.getDomain(uri);
+            var hostname = new URL(uri).hostname;
```
new URL() could fail I imagine. possible null ref?

Hinton replied:
Nice catch, it did indeed throw an exception if the url is invalid.

Hinton commented:
I improved the initial delay and it should no longer jump prior to loading the icon. I decided against adding a default icon since it in my opinion looked worse and more visually distracting.

Chrome also works correctly now, and should the user disable the feature it will instead load the icons cached by the browser. Not sure if there should be an option to completely disable the icons for chrome.

kspearrin commented:
I added response caching to the icons.bitwarden.com endpoint for 24 hours. This should tell the browser to cache the results internally via the `Cache-Control` header. This works as expected on on the web vault and I would think it would work in the browser extension as well (after all its just a DOM window as well). Try it out. It should allow you to just get rid of the internal caching routines altogether if it works.

kspearrin commented:
Looks pretty good in the web vault.

![image](https://user-images.githubusercontent.com/1190944/31418753-f6f47b2e-ae05-11e7-8279-bd4e7a139bf6.png)

Commit if you want to check it out: https://github.com/bitwarden/web/commit/f828288b84fb1fdf3d073cc594a1c6a108fe088c

Hinton commented:
Looks nice. Yea cache headers will probably work and would simplify things a lot.

Hinton commented:
I removed most of the code from the faviconService (caching), and moved the logic out from the current & vault controllers into a component.

As a hack I added the faviconService to config.js since it results in the enabled flag being loaded prior to the vault view and eliminates the late loading.

I think this PR is mostly ready now, the faviconService is more or less only used to keep the config state, not sure if the stateService could be used instead somehow? Which would more or less eliminate it since the last function could be moved into the component.

kspearrin commented on src/popup/app/services/faviconService.js:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        _service.enabled = function () {
+            return enabled || (utilsService.getBrowser() === 'chrome');
```
I think we should do away with the chrome favicons. If someone disables this feature, they may be doing it just because they don't want to see the images at all.

Hinton replied:
True, although it's a nice feature to have for the privacy minded. It's a pity that only chrome supports it.

kspearrin commented on src/popup/app/components/faviconComponent.js:
```diff
@@ -0,0 +1,14 @@
+    	bindings: {
+    		uri: '<'
+    	},
+        template: '<div class="favicon" ng-if="$ctrl.enabled()"><img src="{{$ctrl.url}}"></div>',
```
In a future update I'll be working on very soon in the browser extension we'll be adding "type" icons in here for logins, credit cards, identities, and secure notes. I'll probably want to expand this component for that purpose when that time comes.

Hinton replied:
Should be easy enough, add another binding for the "type" and modify the template slightly.

kspearrin commented on src/popup/app/settings/settingsFeaturesController.js:
```diff
@@ -176,4 +183,28 @@
             });
         };
+
+        $scope.updateDisableFavicon = function () {
```
Do we need to update the application state so when a user navigates back the icons are gone?

kspearrin commented:
Note that I changed the icon path again :)

It is now https://icons.bitwarden.com/{hostname}/icon.png

ex: https://icons.bitwarden.com/bitwarden.com/icon.png

This now allows us to take advantage of our CDN caching the icon. (.ico files are still served, just with a .png extension).

kspearrin commented on src/popup/app/config.js:
```diff
@@ -1,7 +1,9 @@
 
-    .config(function ($stateProvider, $urlRouterProvider, toastrConfig) {
+    .config(function ($stateProvider, $urlRouterProvider, $compileProvider, toastrConfig) {
+        $compileProvider.imgSrcSanitizationWhitelist(/^\s*((moz|chrome)-extension|https:\/\/icons\.bitwarden\.com)/);
```
can be removed?
